### PR TITLE
Backend learning course: first part text document 🚧

### PR DESCRIPTION
**Pull request by gargajit**
Created the text document for first part ie `Basics of http processor and data modelling` of country-details

Added script in planning doc. 

[Rendered Output](https://fastncom-8yvbqyx9m-fifthtry.vercel.app/country-details/basics/)

### CODE REVIEW

File: FASTN.ftd [hunk 1]
```diff
@@ -394,6 +396,13 @@ favicon: /-/fastn.com/images/favicon.svg
 - Dynamic Redirect: /ftd/redirect/
   document: backend/ftd-redirect.ftd
 
+## Learn: /backend/learn/
+  document: /backend/dynamic-ui/index.ftd
+
+- Creating Dynamic UI: /country-details/
+  document: /backend/dynamic-ui/index.ftd
```
Comment (amitu) [inline FASTN.ftd:403]: This file is all wrong. Why is URL country detail but file name dynamic-ui? Everything in backend section is dynamic ui. 

Rename it to backend/examples/country-details.ftd.

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 2]
```diff
@@ -0,0 +1,255 @@
+-- import: fifthtry.github.io/admonitions as cbox
+
+-- ds.page: Basics of `http` processor and `data modelling`
+
+In this part, we will delve into essential concepts like the `http` processor
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:5]:
```suggestion
In this part, we will delve into concepts like the `http` processor
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+and `data modelling` through a concise yet illuminating example.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:6]:
```suggestion
and `data modelling` using a simple example.
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+To facilitate your learning journey, we will take a pratical scenario involving 
+an array of records. Each record has two key attributes: the **country-name** 
+and it's corresponding **capital**, in the form of JSON data.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:10]:
```suggestion
We will  an array of records. Each record has two key attributes: the **country-name** 
and it's corresponding **capital**, in the form of JSON data.
```

Use simpler language. Cut words that are not needed. Avoid adjectives if you can.

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+-- ftd.text: Click to see the JSON data
+link: https://famous-loincloth-ox.cyclic.app/
+open-in-new-tab: true
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:14]: Why are you torturing reader? 

Use `ds.code` and show the JSON right here.

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+`http` processor does the network communication using REST API and fetches data
+from the external site in a different domain.
+
+The syntax of using `http` processor is as follows:
+
+-- ds.code:
+
+\$processor$: pr.http
+
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:40]: This is wrong. Use proper snippet, 

```
-- record r:
$processor$: foo
```

`$processor$` on its own is nonsense.

Comment (amitu) [reply backend/dynamic-ui/http-data-modelling.ftd:40]: Also there is an import statement you have omitted.

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+-- ds.h2: Data modelling: `record`
+
+For this example, we are going to use `fastn`'s native type called `record`.
+`Record` is a user defined type that can be used to hold the country data.
+
+The syntax of record is as follows:
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:47]:
```suggestion
Here we are creating a new record called `country-data`.
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+The syntax of using `http` processor is as follows:
+
+-- ds.code:
+
+\$processor$: pr.http
+
+
+-- ds.h2: Data modelling: `record`
+
+For this example, we are going to use `fastn`'s native type called `record`.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:44]:
```suggestion
For this example, we will use `record` feature of `fastn`'s.
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+`Record` is a user defined type that can be used to hold the country data.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:45]: There is no such thing as `Record`, it is called `record`. 

```suggestion
`record` can be used to create a user defined data type. 
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+The syntax of record is as follows:
+
+-- ds.code: Declaring a country record
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:49]:
```suggestion
-- ds.code: Declaring the `country-record`
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+\-- record country-data:
+string name:
+string capital:
+
+-- ds.h2: Start building
+
+Let's implement the theory and build the project. 
+
+-- ds.h3: **Step 1:** Import `fastn` processor
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:59]:
```suggestion
-- ds.h3: **Step 1:** Import `fastn/processors`
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+string name:
+string capital:
+
+
+-- ds.h3: **Step 3:** Create a list
+
+Since the JSON data has a list of countries and their respective captials, 
+therefore, we will create a list and use the `country-data` record as the type.
+
+-- ds.code: list countries of country type
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:85]:
```suggestion
-- ds.code: `countries` is a `list` of `country-data`
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+\-- country-data list countries:
+
+
+-- ds.markdown:
+
+This list will be created when we do the `http` calling with the `http`
+processor. Also, we will provide the endpoint i.e. the address where the data
+is present as the `url`.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:94]:
```suggestion
`$processor$: pr.http` will will initialise `countries` with the data returned by the `url`. 
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+
+-- ds.code: 
+
+\-- country-data list countries:
+$processor$: pr.http
+url: https://famous-loincloth-ox.cyclic.app/
+
+
+
+-- ds.h3: **Step 4:** Create a component definition
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:104]:
```suggestion
-- ds.h3: **Step 4:** Data is ready, lets show it in the UI
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+
+
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:106]:
```suggestion
Now that we have download the data from the `url` and stored it in `countries`, we want to show it to user. To do that let's create a component called `country-detail`. 


```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+\-- component country-detail:
+
+
+\-- end: country-detail
+
+
+-- ds.markdown: 
+
+Create a header property `country` of type **country-data** as caption
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:117]:
```suggestion
This component will have a property `country`. We will mark it as `caption` to make easy for users of this
component. 
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+-- ds.code:
+
+\-- component country-detail:
+caption country-data country:  ;; <hl>
+
+\-- end: country-detail
+
+-- ds.markdown:
+
+Let's first pull the data of country and display in the browser
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:128]:
```suggestion
Let's show the country name.
```

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+\-- end: country-detail
+
+
+
+
+-- ds.markdown:
+
+Till now, we have just defined the component but to display the list of country
+names we need call the component. Therefore, after closing the component we 
+can call the component and create a `for` loop.
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:146]: You can not create a for loop. Only arpita can create features. You use a for loop. Do you use a chair to sit down or do you create a chair to sit down?

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+
+
+-- ds.code:
+
+\-- country-detail: $country
+for: $country in $countries
+
+-- ds.markdown:
+
+Now wrap the two texts for country-name and capital in `row` container
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:156]: What is `country-name`? I see country.name. And if you call name country-name, then why are you calling capital capital and not country-capital?

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+
+-- ds.markdown:
+
+So you have successfully fetched and displayed the values of JSON data from the
+external website using the `http` processor and one of the data modelling type, 
+`record`.
+
+-- ds.h3: Improved UI
+
+You can use various CSS properties to improve the UI to display the data, let's
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:177]: These are not CSS properties. These are `fastn` properties.

File: backend/dynamic-ui/http-data-modelling.ftd [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+So you have successfully fetched and displayed the values of JSON data from the
+external website using the `http` processor and one of the data modelling type, 
+`record`.
+
+-- ds.h3: Improved UI
+
+You can use various CSS properties to improve the UI to display the data, let's 
+say in the form of a table.
+
+-- ds.code: css properties added
```
Comment (amitu) [inline backend/dynamic-ui/http-data-modelling.ftd:180]: DO not use CSS unless you are using the `css:` property.

File: backend/dynamic-ui/index.ftd [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+
+This guide provides step-by-step instructions to deploy the `country-details-ui`
+project.
+
+
+-- ds.h1: REST API
+
+Application Programming Interface or API is like a set of rules and protocols
+that allow different software applications to talk to each other and share data
+or perform actions.
```
Comment (amitu) [inline backend/dynamic-ui/index.ftd:11]: Delete this para.

Comment (amitu) [reply backend/dynamic-ui/index.ftd:11]: Entire section.

File: backend/dynamic-ui/http-data-modelling.ftd (hunk 5 above)
Comment (gargajit) [reply backend/dynamic-ui/http-data-modelling.ftd:14]: Sure, I have displayed it.

```json
-- ds.code:
lang: json
copy: false

[
  {
    "name": "India",
    "capital": "New Delhi"
  },
  {
    "name": "Sri Lanka",
    "capital": "Colombo"
  },
  {
    "name": "Nepal",
    "capital": "Kathmandu"
  },
  {
    "name": "Bangladesh",
    "capital": "Dhaka"
  },
  {
    "name": "Indonesia",
    "capital": "Jakarta"
  },
  {
    "name": "Maldives",
    "capital": "Mali"
  }
]
```

File: FASTN.ftd (hunk 1 above)
Comment (gargajit) [reply FASTN.ftd:403]: Updated it like this:
```ftd
- Country Details: /country-details/
  document: /backend/country-details/index.ftd
  - Basic of `http`, `data modelling`: /country-details/basics/
    document: /backend/country-details/http-data-modelling.ftd

File: backend/dynamic-ui/http-data-modelling.ftd (hunk 6 above)
Comment (gargajit) [reply backend/dynamic-ui/http-data-modelling.ftd]: Updated 

```
\-- import: fastn/processors as pr  ;; <hl>

\-- record r:
\$processor$: http.pr  ;; <hl>
```

**Pull request merged**